src/rust-crypto/rust-crypto.ts

```diff
@@ -387,42 +387,82 @@ export class RustCrypto implements CryptoBackend {
         createSecretStorageKey,
         setupNewSecretStorage,
     }: CreateSecretStorageOpts = {}): Promise<void> {
-        // If createSecretStorageKey is not set, we stop
-        if (!createSecretStorageKey) return;
+        // If an AES Key is already stored in the secret storage and setupNewSecretStorage is not set
+        // we don't want to create a new key
+        const isNewSecretStorageKeyNeeded = setupNewSecretStorage || !(await this.secretStorageHasAESKey());
 
-        // See if we already have an AES secret-storage key.
-        const secretStorageKeyTuple = await this.secretStorage.getKey();
+        if (isNewSecretStorageKeyNeeded) {
+            if (!createSecretStorageKey) {
+                throw new Error("unable to create a new secret storage key, createSecretStorageKey is not set");
+            }
 
-        if (secretStorageKeyTuple) {
-            const [, keyInfo] = secretStorageKeyTuple;
+            // Create a new storage key and add it to secret storage
+            const recoveryKey = await createSecretStorageKey();
+            await this.addSecretStorageKeyToSecretStorage(recoveryKey);
+        }
 
-            // If an AES Key is already stored in the secret storage and setupNewSecretStorage is not set
-            // we don't want to create a new key
-            if (keyInfo.algorithm === SECRET_STORAGE_ALGORITHM_V1_AES && !setupNewSecretStorage) {
-                return;
+        // If we have cross-signing private keys cached, store them in secret
+        // storage if they are not there already.
+        if (
+            (await this.isCrossSigningReady()) &&
```

Comment:
I think maybe this should call `this.olmMachine.crossSigningStatus()` to check whether we have the private keys directly, rather than `isCrossSigningReady` which makes all sorts of other checks that aren't relevant here.